### ISSUE TITLE
LayerIntegratedGradients does not work with two kinds of inputs

### ISSUE THREAD

**Issue by jungminc88** [pytorch/captum#956]
## 🐛 Bug

My model takes in two inputs, tensors `query` and `article`, the lengths are different but the batch size is the same.
Say, query.size() is [1, 137],  article.size() is [1, 512].
The each input is turned into a tensor of shape [1, 768] by going through BERT and mean pooling.
The model returns the cosine similarity of the two.

I constructed the reference features and tried to compute attributions using LayerIntegratedGradients.
It returns error. I put it in the next section "To Reproduce".
---------------------------------------------------------------------------

## To Reproduce

Here is the minimal (as minimal as I could get it) reproduction of the error

```
from transformers import AutoModel, AutoTokenizer
import torch
import torch.nn as nn
import captum

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
pretrained_name = pretrained_name = "sentence-transformers/paraphrase-distilroberta-base-v1"
tokenizer = AutoTokenizer.from_pretrained(pretrained_name)
model = AutoModel.from_pretrained(
    pretrained_name, # Pre trainedモデルの指定
    output_attentions = False, # アテンションベクトルを出力するか
    output_hidden_states = True, # 隠れ層を出力するか
)
query = "The quick brown fox jumps over the lazy dog"
article = "A fox may jump over a dog"
que_tokenized = tokenizer.encode_plus(query, return_tensors="pt")
que_iis, que_ams = que_tokenized['input_ids'], que_tokenized['attention_mask']
art_tokenized = tokenizer.encode_plus(article, return_tensors="pt")
art_iis, art_ams = art_tokenized['input_ids'], art_tokenized['attention_mask']
que_iis, que_ams, art_iis, art_ams = que_iis.to(device), que_ams.to(device), art_iis.to(device), art_ams.to(device)

from captum.attr import visualization as viz
from captum.attr import LayerConductance, LayerIntegratedGradients
model.cuda()
model.eval()
model.zero_grad()
ref_token_id = tokenizer.pad_token_id # A token used for generating token reference
sep_token_id = tokenizer.sep_token_id # A token used as a separator between question and text and it is also added to the end of the text.
cls_token_id = tokenizer.cls_token_id # A token used for prepending to the concatenated question-text word sequence

def construct_ref_features(que_iis, que_ams, art_iis, art_ams):

    ref_que_iis = torch.where((que_iis==cls_token_id) | (que_iis==sep_token_id), que_iis, ref_token_id)
    ref_art_iis = torch.where((art_iis==cls_token_id) | (art_iis==sep_token_id), art_iis, ref_token_id)


    return ref_que_iis, que_ams, ref_art_iis, art_ams
ref_que_iis, ref_que_ams, ref_art_iis, ref_art_ams = \
  construct_ref_features(que_iis, que_ams, art_iis, art_ams)
def predict(que_iis, art_iis, que_ams, art_ams):

    cos_sim = nn.CosineSimilarity(dim=1, eps=1e-6)
    query_out = model(input_ids=que_iis, attention_mask=que_ams).last_hidden_state
    article_out = model(input_ids=art_iis, attention_mask=art_ams).last_hidden_state
    query_emb = _mean_pooling(query_out, que_ams)
    article_emb = _mean_pooling(article_out, art_ams)
    sim = cos_sim(query_emb, article_emb)

    return sim
def _mean_pooling(model_output_last_hidden_state, attention_mask):
    token_embeddings = model_output_last_hidden_state
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
lig = LayerIntegratedGradients(predict, model.embeddings)

attributions, delta = \
    lig.attribute(inputs=(que_iis, art_iis, que_ams, art_ams), baselines=(ref_que_iis, ref_art_iis, ref_que_ams, ref_art_ams),  return_convergence_delta=True)
```

and the error message
```
RuntimeError                              Traceback (most recent call last)
[<ipython-input-55-0fb4bff40406>](https://localhost:8080/#) in <module>()
----> 1 attributions, delta =     lig.attribute(inputs=(que_iis, art_iis, que_ams, art_ams), baselines=(ref_que_iis, ref_art_iis, ref_que_ams, ref_art_ams),  return_convergence_delta=True)

16 frames
[/usr/local/lib/python3.7/dist-packages/transformers/models/roberta/modeling_roberta.py](https://localhost:8080/#) in forward(self, hidden_states, attention_mask, head_mask, encoder_hidden_states, encoder_attention_mask, past_key_value, output_attentions)
    257         if attention_mask is not None:
    258             # Apply the attention mask is (precomputed for all layers in RobertaModel forward() function)
--> 259             attention_scores = attention_scores + attention_mask
    260 
    261         # Normalize the attention scores to probabilities.

RuntimeError: The size of tensor a (9) must match the size of tensor b (11) at non-singleton dimension 3
```

## Environment
Describe the environment used for Captum

```

 - Captum / PyTorch Version (e.g., 1.0 / 0.4.0): 0.5.0 / 1.11.0+cu113
 - OS (e.g., Linux): Colab
 - How you installed Captum / PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.9
 - CUDA/cuDNN version: cu113
 - GPU models and configuration: Tesla T4
 - Any other relevant information:

## Additional context

<!-- Add any other context about the problem here. -->

Any comment or suggestion is highly appreciated. Thank you in advance!

**Comment by NarineK** [comment]
@jungminc88, IG scales the output of embedding layer and it looks like the other matrix that we concatenate with is not scaled.  My guess is that one of these tensors `attention_scores, attention_mask` is be scaled and the other one not.
https://github.com/huggingface/transformers/blob/9fc34235fa3329c918d5ba67ce09a0cc8f399c59/src/transformers/models/roberta/modeling_roberta.py#L749
Perhaps we can use `configure_interpretable_embedding_layer` and IntegratedGradients instead of the LayerIG. That will help us to scale the mask as well.

Here is an example: https://captum.ai/tutorials/Bert_SQUAD_Interpret2

Basically, we want that the inputs of our model are embeddings not the token indices because otherwise we cannot compute the gradients w.r.t. indices.

**Comment by craymichael** [comment]
Hi, this is covered by the BERT tutorials and FAQ. For text models, attribute to embeddings rather than token IDs; for multiple text inputs, use a wrapper and pass aligned tensor inputs / additional args so the scalar similarity score is the attribution objective. Closing as documented. https://captum.ai/tutorials/Bert_SQUAD_Interpret2 https://captum.ai/docs/faq
